tinygrad/codegen/linearizer.py

```diff
@@ -479,15 +480,19 @@ def uop(self, uop:UOps, dtype:Optional[DType]=None, vin:Tuple[UOp, ...]=tuple(),
     key = (uop, dtype, vin, arg)
 
     if uop == UOps.ALU:
-      if arg == BinaryOps.CMPLT: assert dtype == dtypes.bool, f"{arg} output dtype mismatch {dtype=} != {dtypes.bool}"
-      if arg == TernaryOps.WHERE: assert vin[0].dtype == dtypes.bool, f"{arg} selector dtype mismatch {vin[0].dtype=} != {dtypes.bool}"
-
-    if uop == UOps.PHI and vin[1].dtype != dtype: vin = (vin[0], self.cast(vin[1], dtype)) + vin[1:]
-    if uop == UOps.ALU: # upcast vins to the same dtype
-      upcast_dtype = dtypes.float if arg == TernaryOps.MULACC else max(cast(DType, x.dtype) for x in vin) # MULACC is only supported in float
-      if arg == TernaryOps.WHERE: vin = (vin[0],) + tuple(self.cast(x, upcast_dtype) for x in vin[1:]) # the first arg is always bool
-      else: vin = tuple(self.cast(x, upcast_dtype) for x in vin)
-      dtype = dtype or upcast_dtype # some ops like BinaryOps.CMPLT return bool
+      if arg in UnaryOps:
+        assert dtype == vin[0].dtype, f"{arg} dtype mismatch {dtype=} != {vin[0].dtype=}"
+      elif arg == BinaryOps.CMPLT:
+        assert dtype == dtypes.bool, f"{arg} output dtype mismatch {vin[0].dtype=} != {dtypes.bool}"
+        assert vin[0].dtype == vin[1].dtype, f"{arg} dtype mismatch {dtype=} != {vin[0].dtype=} != {vin[1].dtype=}"
+      elif arg in BinaryOps:
+        assert dtype == vin[0].dtype == vin[1].dtype, f"{arg} dtype mismatch {dtype=} != {vin[0].dtype=} != {vin[1].dtype=} {vin[0]} {vin[1]}"
+      elif arg == TernaryOps.WHERE:
+        assert vin[0].dtype == dtypes.bool, f"{arg} selector dtype mismatch {vin[0].dtype=} != {dtypes.bool}"
+        assert dtype == vin[1].dtype == vin[2].dtype, f"{arg} choice dtype mismatch {dtype=} != {vin[1].dtype=} != {vin[2].dtype=}"
+
+    # noqa: E501 if uop == UOps.PHI: assert vin[1].dtype == vin[0].dtype, f"PHI dtype mismatch {dtype=} != {vin[0].dtype=} != {vin[1].dtype=}" TODO this is a wmma bug L317
```

Comment:
can you elaborate more on this or file an issue? also it should not need `noqa: E501` and comments are not counted in code size

Reply:
do you think the assert is wrong?
it fails in wmma because of [op3's dtype](https://github.com/tinygrad/tinygrad/blob/master/tinygrad/codegen/linearizer.py#L314C63-L314C63)
there's a chance I'm wrong here and wmma is behaving as expected though